src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,562 @@
+        tabular.record('{}/KLBefore'.format(self.policy.name),
+                       policy_kl_before)
+        tabular.record('{}/KL'.format(self.policy.name), policy_kl)
+        pol_ent = self._f_policy_entropy(*policy_opt_input_values)
+        ent = np.sum(pol_ent) / np.sum(samples_data['valids'])
+        tabular.record('{}/Entropy'.format(self.policy.name), ent)
+        tabular.record('{}/Perplexity'.format(self.policy.name), np.exp(ent))
+
+        self._fit_baseline_with_data(samples_data)
+
+        ev = np_tensor_utils.explained_variance_1d(samples_data['baselines'],
+                                                   samples_data['returns'],
+                                                   samples_data['valids'])
+
+        tabular.record('{}/ExplainedVariance'.format(self.baseline.name), ev)
+        self._old_policy.model.parameters = self.policy.model.parameters
+
+    def _build_inputs(self):
+        """Build input variables.
+
+        Returns:
+            namedtuple: Collection of variables to compute policy loss.
+            namedtuple: Collection of variables to do policy optimization.
+
+        """
+        observation_space = self.policy.observation_space
+        action_space = self.policy.action_space
+
+        with tf.name_scope('inputs'):
+            if self.flatten_input:
+                obs_var = tf.compat.v1.placeholder(
+                    tf.float32,
+                    shape=[None, None, observation_space.flat_dim],
+                    name='obs')
+            else:
+                obs_var = observation_space.to_tf_placeholder(name='obs',
+                                                              batch_dims=2)
+            action_var = action_space.to_tf_placeholder(name='action',
+                                                        batch_dims=2)
+            reward_var = new_tensor(name='reward', ndim=2, dtype=tf.float32)
+            valid_var = tf.compat.v1.placeholder(tf.float32,
+                                                 shape=[None, None],
+                                                 name='valid')
+            length_var = new_tensor(name='length', ndim=1, dtype=tf.int32)
+            baseline_var = new_tensor(name='baseline',
+                                      ndim=2,
+                                      dtype=tf.float32)
+
+            policy_state_info_vars = {
+                k: tf.compat.v1.placeholder(tf.float32,
+                                            shape=[None] * 2 + list(shape),
+                                            name=k)
+                for k, shape in self.policy.state_info_specs
+            }
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k] for k in self.policy.state_info_keys
+            ]
+
+        if self.policy.recurrent:
+            _obs_var = obs_var
+            # pylint: disable= protected-access
+            if self.policy._state_include_action:
```

Comment:
i don't see the relationship here -- you test `_state_include_action` and then indiscriminately augment the input with all of the state_info_vars

Reply:
looks like the comment above is for this block of comment? Yes if we have empty `self.policy.state_info_keys` for non-recurrent policies, we can always augment with `self.policy.state_info_keys`.

Reply:
this is what i'm proposing. i want to delete the property `Policy.recurrent`